### DISCUSSION TITLE
openDTU und MQTT

### DISCUSSION THREAD

**Discussion by frontlinerde** [tbnobody/OpenDTU#525] [Q&A]
Hi, versuche meinen HM-600 und HM-1500 per iobroker und MQTT zusteuern.
Hauptsächlich geht es mir um "limit_nonpersistent_absolute" und "limit_persistent_absolute"
habe die Datenpunkte unter mqtt.0.solar.xxx.cmd angelegt.
Wenn ich die Werte versuche zu ändern, passiert nichts, der WR übernimmt diese nicht.

Muss noch irgendwas aktiviert werden?

Danke

**Comment by JauntyJosef** [comment]
Ich nehme mal an, dass anstatt xxx natürlich die Seriennummer des WR steht.
Benutzt den ioBroker MQTT Adapter als Server oder Client?

**Reply by frontlinerde** [reply]
ja mit xxx ist die Serienummer gemeint.
Der Adapter läuft als Server.

Denke ich habs aber gefunden, "Nachrichten ohne "retain"-Flag senden" war nicht aktiviert.

seither gehts, muss jetzt nur schauen was die anderen Geräte die auf dieser Instanz laufen nun machen.

**Reply by JauntyJosef** [reply]
Ich kann aus eigener Erfahrung sagen, dass der ioBroker Adapter als Server bei mir in ganz vielen Fällen oft unerklärliche Probleme machte (auch mit retained messages) Da als Basissystem bei mir ohnehin Linux läuft, habe ich einfach den mosquitto Broker installiert und die Adapter Instanz im ioBroker, die als Server diente, auf Client umgestellt. Dadurch musste ich bei allen anderen MQTT Clients nichts an der Konfiguration ändern, da die IP des Servers gleicht blieb und die Datenpunkte im ioBroker konnte ich auch 1:1 weiter verwenden.

**Comment by RichieXX** [comment]
@frontlinerde 
Nur das Anlegen des Datenpunktes in iob hat bei mir auch nicht funktioniert. Ich habe den Datenpunkt mit dem MQTT Explorer "published" und dann noch den Typ auf "Zahl"/"number" geändert. Damit haben meine bisherigen Tests funktioniert, hatte aber noch nicht viel damit zu tun (das Wetter ist zu schlecht zum limitieren).

**Comment by BlackandGold1** [comment]
Gibt es da noch Tricks und Tipps. Bei mir geht da gar nix. Kann die Werte ändern wie ich will , es geht nichts???? 
Hoymiles HM 1500 mit openDTU verbunden und läuft. Im Iobroker mit Mqtt Broker Adapter verbunden Daten vorhanden: Datenpunkt unter mqtt.0.solar.xxxxxxx.cmd.limit_nonpersistent_absolute angelegt mit Mqtt explorer ,erscheint auch. Ja und jetzt??? Egal wie ich den Wert ändere es passiert nichts.

**Reply by frontlinerde** [reply]
Werte müssen ohne "retain"-Flag gesendet werden

**Reply by BlackandGold1** [reply]
> Werte müssen ohne "retain"-Flag gesendet werden

Sorry wenn ich jetzt dumm frage. Aber das heißt was genau.?

**Comment by BlackandGold1** [comment]
Nachtrag für die Nachwelt: Ich musste alle Haken die Standard mässig im Iobroker Adapter Mqtt Broker drin sind, entfernen. Lediglich der Haken bei:" ohne retain Flag senden" muss drin sein. Konnte dann die Datenpunkte im Expertenmodus hinzufügen und kann sie nun ganz normal steuern. WICHTIG: In Blockly immer den Steuer Block nehmen, der Wert darf nicht bestätigt werden.

**Comment by jimknopf63** [comment]
Im Anhang ein kleines Java Script, wie es bei mir läuft um eine nahezu Nulleinspeisung zu realisieren. Eigentlich ziemlich Selbsterklärend. Daten vom Smartmeter kommen über einen Hitchi IR Lesekopf.

<img width="849" alt="Bildschirm­foto 2023-02-03 um 17 34 36" src="https://user-images.githubusercontent.com/97401894/218777225-06a0f229-2134-4354-9301-f193327f229b.png">

**Reply by BlackandGold1** [reply]
> Im Anhang ein kleines Java Script, wie es bei mir läuft um eine nahezu Nulleinspeisung zu realisieren. Eigentlich ziemlich selbsterklärend. Daten vom Smartmeter kommen über einen Hitchi IR Lesekopf.
> 
> <img alt="Bildschirmfoto 2023-02-03 um 17 34 36" width="849" src="https://user-images.githubusercontent.com/97401894/218777225-06a0f229-2134-4354-9301-f193327f229b.png">

Danke für diesen Script.Wie läuft es bei dir? Bei mir kommen unmöglich Wert raus. Entweder ich Beziehe oder Speise ein.. Was gibt bei dir der Datenpunkt gridSum an?

**Reply by BlackandGold1** [reply]
> > Im Anhang ein kleines Java Script, wie es bei mir läuft um eine nahezu Nulleinspeisung zu realisieren. Eigentlich ziemlich selbsterklärend. Daten vom Smartmeter kommen über einen Hitchi IR Lesekopf.
> > <img alt="Bildschirmfoto 2023-02-03 um 17 34 36" width="849" src="https://user-images.githubusercontent.com/97401894/218777225-06a0f229-2134-4354-9301-f193327f229b.png">
> 
> Danke für dieses Script.Wie läuft es bei dir? Bei mir kommen unmögliche Wert raus. Entweder ich Beziehe oder Speise ein.. Was gibt bei dir der Datenpunkt gridSum an?

**Reply by frontlinerde** [reply]
gridSum gibt den aktuellen Verbrauch an

**Reply by BlackandGold1** [reply]
> gridSum gibt den aktuellen Verbrauch an

Okay.Dann hab ich alles richtig gemacht. Leider geht es trotzdem nicht..Er regelt wild hoch und runter.

**Reply by jimknopf63** [reply]
> > gridSum gibt den aktuellen Verbrauch an
> 
> Okay.Dann hab ich alles richtig gemacht. Leider geht es trotzdem nicht..Er regelt wild hoch und runter.

Da schwankt dein Verbrauch gewaltig, hast Du auch deine Grundlast eingetragen. Bei mir 200. Mit minDifference kannst Du den Wert einstellen der mindestens vorhanden sein muss um eine erneute Änderung auszulösen. Ich habe da 30 stehen bei mir

**Reply by BlackandGold1** [reply]
> > > gridSum gibt den aktuellen Verbrauch an
> > 
> > 
> > Okay.Dann hab ich alles richtig gemacht. Leider geht es trotzdem nicht..Er regelt wild hoch und runter.
> 
> Da schwankt dein Verbrauch gewaltig, hast du auch deine Grundlast eingetragen. Bei mir 200. Mit minDifference kannst Du den Wert einstellen der mindestens vorhanden sein muss um eine Änderung auszulösen. Ich habe da 30 stehen bei mir

Ja habe 200 und 150  bei Grundlast probiert und  auch mit Differenzen habe ich gespielt. Habe die Zeit auch auf 9 Sekunden probiert da der Hoymiles gar nicht hinterherkommt mit regeln. Meine Grundlast ist ziemlich gleich da nur Router und Qnap Nas laufen wenn keiner daheim ist. Zusätzlich springt natürliche der Kühlschrank alle paar Stunden an mit 80 Watt. Ich teste weiter.

**Comment by rebel1985** [comment]
hallo,
das skript habe ich auch nur habe ich ein problem, es kommt wohl mit negativ werten vom stromzähler nicht klar.
kann man das irgendwie abfangen oder umgehen?
tags über läuft noch ein kleines Balkonsolar mit und der akku soll hal dazu regeln und aufladen damit er nachts einspeisen kann.
```
let ids = {
    //Stromzähler
    gridSum: 'sonoff.0.Smartzaehler-Strom.SML_curr_w',    
    //WR_Steuerung
    solarLimit: 'opendtu.0.112182854324.power_control.limit_nonpersistent_absolute', 
    //           opendtu.0.114184519980.power_control.limit_nonpersistent_absolute                
    //WR_Trennung_und_Power_Messung
    solarInput: 'shelly.0.shellyplus1pm#a8032ab78318#1.Relay0.Power',
    }

let inverterOutput = {
    min: 10, // Minimales Limit (Grundverbrauch)
    max: 100, // Maximales Limit (Wechselrichter Leistung max.)
    safetyMargin: -55, // Einspeisungsziel (da sonst kleine Anstiege im Stromverbrauch auf das Stromnetz gehen)
    minDifference: 15, // Kein neues Limit setzen, wenn Differenz alt neu <= dieser Wert ist
}

let oldLimit = 0;

// Run every 15 seconds
schedule('*/15 * * * * *', function () {
    let fromSolar = getState(ids.solarInput).val;
    let fromGrid = getState(ids.gridSum).val;

    let newLimit = fromSolar + fromGrid + inverterOutput.safetyMargin;
    if (newLimit < inverterOutput.min) newLimit = inverterOutput.min;
    if (newLimit > inverterOutput.max) newLimit = inverterOutput.max;
    if (inverterOutput.min > fromSolar) newLimit = inverterOutput.max;

    if(!(Math.abs(newLimit - oldLimit) <= inverterOutput.minDifference))
    {
        log(`Setting limit to ${fromSolar} + ${fromGrid} = ${newLimit}W`);
        setState (ids.solarLimit , newLimit);
        oldLimit = newLimit;
    }
});
```

**Reply by jimknopf63** [reply]
Wenn das von meinem Script abgewandelt ist dann darfst Du bei saftyMargin: nur 55 eingeben. Das - gehört da nicht hin weil bei dem Wert von einer Einspeisung von 55 W ausgegangen wird.

**Reply by rebel1985** [reply]
hi,
könntes du mir vllt. dann mal bitte dein Skipt zu kommen lassen was bei dir läuft?
ich habe es nur in einem forum über pn bekommen.

`<html><body>


17:43:10.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 132 - 40 = 94W
-- | -- | --
17:43:20.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 92.7 - 40 = 54.7W
17:43:30.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 53.8 - 40 = 15.799999999999997W
17:43:40.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 30 + 10.7 - 40 = 7W
17:43:50.000 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 31 + 5.7 - 40 = 200W
17:44:00.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 208.2 - 40 = 170.2W
17:44:10.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 171.1 - 40 = 133.1W
17:44:20.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 133.2 - 40 = 95.19999999999999W
17:44:30.001 | info | javascript.0 (1393587) script.js.Zu_Hause.PV_AKKU.WR_Akku_Steuerung_Nachts: Setting limit to 2 + 0 - 40 = 200W


</body>
</html>`

**Reply by jimknopf63** [reply]
Mein Script habe ich hier im Verlauf gepostet, steht 4-5 Beiträge weiter oben (14. Feb)

[OpenDTU Nulleinspeisung.txt](https://github.com/tbnobody/OpenDTU/files/12324039/OpenDTU.Nulleinspeisung.txt)

**Reply by rebel1985** [reply]
```
// Run every 10 seconds
schedule('*/10 * * * * *', function () {
    let fromSolar = getState(ids.solarInput).val;
    let oldLimit = getState(ids.solarLimit).val;
    let fromGrid = getState(ids.gridSum).val;
    if (ids.gridSum < 2) fromGrid = 2;
    
    let newLimit = fromGrid + fromSolar - inverterOutput.safetyMargin;
    if (fromGrid < 2) newLimit = inverterOutput.min;
    if (newLimit < inverterOutput.min) newLimit = inverterOutput.min;
    if (newLimit > inverterOutput.max) newLimit = inverterOutput.max;
    //if (inverterOutput.min > fromSolar) newLimit = inverterOutput.max;
```

wozu hast du diese berrechnung eingefügt?
if (inverterOutput.min > fromSolar) newLimit = inverterOutput.max;
bei mir musste ich die ausschalten wegen dem negativ wert.

**Comment by rebel1985** [comment]
das habe ich jetzt soweit zum laufen bekommen

```
let ids = {
    //Stromzähler
    gridSum: 'sonoff.0.Smartzaehler-Strom.SML_curr_w',    
    //WR_Steuerung
    solarLimit: 'opendtu.0.112182854324.power_control.limit_nonpersistent_absolute', 
    //           opendtu.0.114184519980.power_control.limit_nonpersistent_absolute                
    //WR_Trennung_und_Power_Messung
    solarInput: 'shelly.0.shellyplus1pm#a8032ab78318#1.Relay0.Power',
    }

let inverterOutput = {
    min: 7, // Minimales Limit (Grundverbrauch)
    max: 210, // Maximales Limit (Wechselrichter Leistung max.)
    safetyMargin: 40, // Einspeisungsziel (da sonst kleine Anstiege im Stromverbrauch auf das Stromnetz gehen)
    minDifference: 5, // Kein neues Limit setzen, wenn Differenz alt neu <= dieser Wert ist
}

//let oldLimit = 0;

// Run every 5 seconds
schedule('*/5 * * * * *', function () {
    let fromSolar = getState(ids.solarInput).val;
    let oldLimit = getState(ids.solarLimit).val;
    let fromGrid = getState(ids.gridSum).val;
    if (getState(ids.gridSum).val < 5) fromGrid = 4;
    //if (ids.gridSum < 2) fromGrid = 2;
    
    let newLimit = fromGrid + fromSolar - inverterOutput.safetyMargin;
    if (fromGrid < 10) newLimit = inverterOutput.min;
    if (newLimit < inverterOutput.min) newLimit = inverterOutput.min;
    if (newLimit > inverterOutput.max) newLimit = inverterOutput.max;
    //if (inverterOutput.min > fromSolar) newLimit = inverterOutput.max;

    if(!(Math.abs(newLimit - oldLimit) <= inverterOutput.minDifference))
    {
        log(`Setting limit to ${fromGrid} + ${fromSolar} - ${inverterOutput.safetyMargin} = ${newLimit}W`);
        setState (ids.solarLimit , newLimit);
        oldLimit = newLimit;
    }
});

//OLD
//    let fromSolar = getState(ids.solarInput).val;
//    let fromGrid = getState(ids.gridSum).val;

//    let newLimit = fromSolar + fromGrid - inverterOutput.safetyMargin;
//    if (fromGrid < +1) newLimit = inverterOutput.min;
//    if (newLimit < inverterOutput.min) newLimit = inverterOutput.min;
//    if (newLimit > inverterOutput.max) newLimit = inverterOutput.max;
//    if (inverterOutput.min > fromSolar) newLimit = inverterOutput.max;
```
